charm/ksql/reactive/ksql.py

```diff
@@ -13,62 +13,40 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-import os
-import socket
-import tempfile
-import glob
-import subprocess
+from charms.layer.ksql import Ksql, KSQL_SNAP_COMMON
 
-from pathlib import Path
-from OpenSSL import crypto
-from subprocess import check_call
-
-from charms.reactive import set_state, remove_state, when, when_not, hook, when_file_changed
 from charmhelpers.core import hookenv
+
+from charms.reactive import (set_state, remove_state, when, when_not,
+                             hook, clear_flag, is_flag_set, set_flag,
+                             when_any)
 from charms.reactive.helpers import data_changed
-from charmhelpers.core.hookenv import log, config
 
-from charms.layer.ksql import KSQL_SNAP, KSQL_SNAP_COMMON, KSQL_PORT, Ksql
-from charms.layer import tls_client
+from charmhelpers.core.hookenv import log
 
-@when_not('ksql.available')
-def install():
-    install_snap()
 
-@hook('upgrade-charm')
-def upgrade():
-    upgrade_snap()
+@hook('config-changed')
+def config_changed():
+    remove_state('ksql.configured')
+    set_flag('ksql.force-reconfigure')
 
-@hook('stop')
-def uninstall():
-    try:
-        check_call(['snap', 'remove', KSQL_SNAP])
-    except subprocess.CalledProcessError as e:
-        hookenv.log("failed to remove snap: {}".format(e))
 
-def install_snap():
-    # Need to install the core snap explicit. If not, there's
-    # no slots for removable-media on a bionic install.
-    # Not sure if that's a snapd bug or intended behavior.
-    check_call(['snap', 'install', 'core'])
+@when('snap.installed.ksql-server')
+@when_not('ksql.configured')
+def configure():
+    ksql = Ksql()
+    kafkas = ksql.get_kafkas()
 
-    cfg = config()
-    # KSQL-Server's snap presedence is:
-    # 1. Included in snap
-    # 2. Included with resource of charm of 'ksql-server'
-    # 3. Snap store with release channel specified in config
-    snap_file = get_snap_file_from_charm() or hookenv.resource_get('ksql-server')
-    if snap_file:
-        check_call(['snap', 'install', '--dangerous', snap_file])
-    if not snap_file:
-        check_call(['snap', 'install', '--{}'.format(cfg['ksql-release-channel']), KSQL_SNAP])
-    
-    set_state('ksql.available')
+    if data_changed('ksql.kafka_units', kafkas) or is_flag_set('ksql.force-reconfigure'):
```

Comment:
What if ksql-cluster-id changes?

Reply:
Or, tls settings the tls relations...?

Reply:
@cmars For the TLS, tls_client.certs.changed would get fired I would think.

Reply:
> What if ksql-cluster-id changes?

@cmars 'config-changed' hook will be fired causing a force reconfigure.

Reply:
@kngu9 Ok.. I think then we need to add ksql-cluster-id to the data_changed conditional here. Otherwise we'll execute configure, but never get past this conditional.

Reply:
@kngu9 Nevermind.. force-reconfigure.. it's fine :)